File: tools/src/main/java/vip/mystery0/rx/PackageDataObserver.java
package vip.mystery0.rx;

import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;

public interface PackageDataObserver<T> extends Observer<PackageData<T>> {
	@Override
	default void onChanged(PackageData<T> tPackageData) {
		switch (tPackageData.getStatus()) {
			case Loading:
				loading();
				break;
			case Content:
				content(tPackageData.getData());
				break;
			case Empty:
				empty(tPackageData.getData());
				break;
			case Error:
				error(tPackageData.getData(), tPackageData.getError());
				break;
			default:
				break;
		}
	}

	default void loading() {
	}

	default void empty(@Nullable T data) {
	}

	default void content(@Nullable T data) {
	}

	default void error(@Nullable T data, @Nullable Throwable e) {
	}
}

File: tools/src/main/java/vip/mystery0/rx/PackageDataObserver.kt
package vip.mystery0.rx

import androidx.lifecycle.Observer

interface PackageDataObserver<T> : Observer<PackageData<T>> {
	override fun onChanged(tPackageData: PackageData<T>) {
		when (tPackageData.status) {
			Status.Loading -> loading()
			Status.Content -> {
				content(tPackageData.data)
				tPackageData.data?.let { contentNoEmpty(it) }
			}
			Status.Empty -> {
				empty(tPackageData.data)
				empty()
			}
			Status.Error -> {
				error(tPackageData.data, tPackageData.error)
				tPackageData.data?.let { error(tPackageData.error) }
			}
		}
	}

	fun loading() {}

	fun empty(data: T?) {}

	fun empty() {}

	fun content(data: T?) {}

	fun contentNoEmpty(data: T) {}

	fun error(e: Throwable?) {}

	fun error(data: T?, e: Throwable?) {}
}
